### ISSUE TITLE
Need monitoring for clusters

### ISSUE THREAD

**Issue by thockin** [kubernetes/k8s.io#401]
We now have `aaa` up as a cluster but we have not set up monitoring or alerting.

We need someone to own and drive this topic before we get too too far with the stuff we run in there.

**Comment by gianrubio** [comment]
I’d be glad to work on this topic, is there a proposal for this topic? What are the requirements for monitoring and alerting? 

I used to be the maintainer of Prometheus operator helm chart, I had quite some experience on monitoring and alerting on kubernetes.

**Comment by thockin** [comment]
Context is that we have a GKE cluster.  We are starting to load-in things
to that cluster.  as we do so, we need to explore what the minimum
requirements for monitoring ought to be, what service to use (stackdriver
seems obvious, but I am open), how to configure that programmatically, and
what our norms will be for alerting and pagers.

**Comment by aditya-konarde** [comment]
/me subscribes

I landed here from @thockin 's tweet looking for volunteers in the infra group. I'm also willing to help with this issue :)

**Comment by stretchcloud** [comment]
Coming here from @thockin tweet. Would love to step in and help.

**Comment by vielmetti** [comment]
For reference, the discussion of standing up `aaa` is in https://github.com/kubernetes/k8s.io/pull/365

**Comment by jakirpatel** [comment]
Hello , 

This is Jakir. Currently I am working on monitoring and alerting system for an organization. So I ll be happy to contribute and maintain it. 

I have few ideas and scenarios for monitoring and alerting. Please let me know how to contribute. 

Jakir.

**Comment by thockin** [comment]
OK, so we have a GKE cluster called "aaa".  It is managed by terraform code checked in here (yay).

I think we have 2 related (but not identical) things to tackle:

1) Cluster monitoring
  a) What should we use?
  b) How do we set it up with git-ops?
  c) What exactly are we concerned about (signals)?
  d) How are alerts delivered to a group of people?
  e) How do we manage that group?
  f) Do we need an on-call rotation?

2) App monitoring
  a) Same tool as cluster monitoring?
  b) What is the minimum expectation for an app to be deployed into community space
  c) How do we manage groups of alertees for each app (ggroups?)
  d) How do we manage on-call for each app?

We have, more or less, nothing established.  We have very rudimentary alerts for k8s.io going down and not much else.  While I lean towards StackDriver (managed solutions FTW) I am going into this with an open mind.  I do not personally have much experience in setting these things up and almost no experience with what the rest of the world does here (Google has our own universe).

What I really want to get is a proposal considering concrete approaches and tradeoffs.

**Comment by jakirpatel** [comment]
For monitoring and alerting, 

Cluster Monitoring : 
a) What should we use?
1. Combination of the prometheus and alertmanager is the good approach. For cluster monitoring kubernetes API's already exposes the metrics. For HA we can just use the solutions like Thanos which is reliable and HA.

b) How do we set it up with git-ops?
We should connect the git-ops (with API's) when some event is triggered. 

c) What exactly are we concerned about (signals)?
Average Response Time of our different components of the cluster. However we should have a distribution of the requests in quantiles. Defining the SLA's for each component will be useful.

d) How are alerts delivered to a group of people?
Mailing lists are useful to deliver the alerts. We can trigger the pager with ML too. 

e)How do we manage that group?
If we have proper set of permission with mailing list we can add those permissions to our alerting system too. 

f) Do we need an on-call rotation?
It will be useful as there are people working all over the world. 

Useful Links: 
https://github.com/thanos-io/thanos
https://prometheus.io/


This will be applicable for the app monitoring too.  If each app exposes the useful metrics then we should give them freedom to setting up alerts and pagers.

**Comment by rajibmitra** [comment]
I would like to see a combination of stackdriver and grafana.

**Comment by RiRa12621** [comment]
I am very much in favor of what @jakirpatel is proposing but would probably go a little further and propose the usage of Prometheus Operator.
I think using the prometheus operator would allow us to follow a git ops approach a little more:
Everything that live on the clusters, like apps of any kind can be following a structure of what they need to deploy in order to enable monitoring and alerting or the app.
So all that is needed for an app to have monitoring/alerting are `promtheusrules` and `servicemonitor` or `podmonitor`, basically enabling monitring alongside the app deployment.

I am happy to help with drafting and implementation approach and actually implementing a solution like that.

**Comment by jmhardison** [comment]
While we are using a slew of tools ourselves, some interesting tooling that has come up lately is Jaeger. However that aside, a focus on @thockin 1C would be good.

Flesh out the requirements (Hard requirements, nice-to-have's, etc) and then source a tool(s) to fit those requirements. Monitoring infrastructure alone can have different requirements from application (end-to-end) user tracing and monitoring.

Example thought pattern:

- Required - Consolidated logging of all related components.
- Required - full client transaction tracing across independent micro-services.
- Optional - Use of PaaS or SaaS offerings in place of managed components where possible.
- Required - Monitoring platform (user access/visibility) is external to the cluster being monitored, or deployed using components that won't be impacted by a cluster outage.
- Required - Native/existing deployment capability with Terraform.
- Required - Non-googlers can access underlying data and/or dashboards.

Fundamentally, what we look at is merging these two bases (app and inf) to have the holistic view of the environment. (As noted by everyone.) Where possible we also choose to adopt PaaS and SaaS offerings to reduce the operational burden and technical debt we are already dealing with. Usually coinciding with the desire or statement that "we are not in the business of running a monitoring tool" but rather we focus our efforts on the service being delivered, not the tooling doing the monitoring. And subsequently drives our desire to leverage cloud native tooling first before adding a tool. (IE, does the cloud provider offer a PaaS/SaaS that has coverage for our requirements.)

With that said, if in a position to use it, we would absolutely use something like Stackdriver where it meets all of our functional requirements. I know on the surface for me it does tick boxes on both the infrastructure monitoring and application tracing side of the fence. And gives the added benefit of neither being inside of the platform being monitored and alerted on, as well as not increasing ops burden as another component that requires infrastructure management.

**Comment by sgmac** [comment]
What is the status of the monitoring/alerting? Is already a defined workgroup with the goals and definition of the architecture? It would be good to agree at least on the definition, and start subdividing tasks. WDYT?

**Comment by jakirpatel** [comment]
@sgmac  agree with you. Lets have this discussion in the weekly meeting I think.

**Comment by RiRa12621** [comment]
I did a draft diagram for using the Prometheus Operator[PrometheusOperator.pdf](https://github.com/kubernetes/k8s.io/files/3840601/PrometheusOperator.pdf)

The idea is that everything in a `monitoring` namespace is deployed from within a shared repository along with all needed RBAC.
Afterwards subgroups can be fully self contained and just deployed anything new they need to have monitored to their namespace and it will be picked up by the prometheus operator and added to their prometheus.

In an example scenario `Team A` decides to create a new app that is running in their namespace on a cluster. Then the team just deploys the app, a Service, ServiceMonitor, PrometheusRule to their namespace and the pods will be scraped by their prometheus and monitored accordingly with their rules.
Basically this allows for each subgroup to have full control over their share of the monitoring.
This also allows to reduce the workload on each individual at some point carrying the pager as they'd only be alerted for parts that they're involved with and not *everything*.
This follows the principle of "only alert who can action an alert".
Also due to the ownership of alerting directly with the sub-groups this allows for fast movement.

Aside all of that, I believe it makes sense for us to actually leverage opensource solutions and with Prometheus a CNCF project rather than google cloud functionality where possible.

**Comment by jakirpatel** [comment]
Thank you @RiRa12621. 
Self service prometheus will be more good approach where the app teams can use the API's to do operations. 

Agree it make sense using the CNCF graduated project like prometheus. It will be more cleaner approach considering app and cluster monitoring.

**Comment by scottilee** [comment]
Hello, I created a doc https://docs.google.com/document/d/1MqGDxrQjKoEXa2zuXwX9dpe1EY-Tb3hRInyjOYKy2iA/edit?usp=sharing and would appreciate any feedback, especially from @thockin!

**Comment by bartsmykla** [comment]
@scottilee it's great to see you started working on the topic! :-) Good timing because I want to start talking about it more during the next weeks. If you'll need any help feel free to reach to me.
